aiidalab_widgets_base/computational_resources.py

```diff
@@ -375,126 +437,131 @@ def _write_ssh_config(self, private_key_abs_fname=None):
+            # unwrap private key file and setting temporary private_key content
+            private_key_abs_fname, private_key_content = self._private_key
+            if private_key_abs_fname is None:  # check private key file
+                self.message = "Please upload your private key file."
+                return False
+
+            # Write private key in ~/.ssh/ and use the name of upload file,
+            # if exist, generate random string and append to filename then override current name.
+            self._add_private_key(private_key_abs_fname, private_key_content)
+
+        if not self._is_in_config():
+            self._write_ssh_config(private_key_abs_fname=private_key_abs_fname)
+
+        # Copy public key on the remote computer.
+        ssh_connection_thread = threading.Thread(target=self._ssh_copy_id)
+        ssh_connection_thread.start()
+
+    def _ssh_copy_id(self):
+        """Run the ssh-copy-id command and follow it until it is completed."""
+        timeout = 30
+        self.password_message = f"Sending public key to {self.hostname.value}... "
+        self._ssh_connection_process = pexpect.spawn(
+            f"ssh-copy-id {self.hostname.value}"
+        )
+        while True:
             try:
-                ssh_copy_id()
-            except StopIteration:
-                print(f"Unsuccessful attempt to connect to {self.hostname.value}.")
-                return
+                self.ssh_connection_state = self._ssh_connection_process.expect(
+                    self.SSH_POSSIBLE_RESPONSES,
+                    timeout=timeout,
+                )
+            except pexpect.TIMEOUT:
+                self._ssh_password.disabled = True
+                self._continue_with_password_button.disabled = True
+                self.password_message = (
+                    f"Exceeded {timeout} s timeout. Please start again."
+                )
+                break
+
+            # Terminating the process when nothing else can be done.
+            if self.ssh_connection_state in (
+                SshConnectionState.success,
+                SshConnectionState.keys_already_present,
+                SshConnectionState.no_keys,
+                SshConnectionState.unknown_hostname,
+                SshConnectionState.connection_refused,
+                SshConnectionState.end_of_file,
+            ):
+                break
+
+        self._ssh_connection_message = None
+        self._ssh_connection_process = None
+
+    def _send_password(self, _=None):
+        self._ssh_password.disabled = True
+        self._continue_with_password_button.disabled = True
+        self._ssh_connection_process.sendline(self._ssh_password.value)
+
+    @traitlets.observe("ssh_connection_state")
+    def _observe_ssh_connnection_state(self, _=None):
+        """Observe the ssh connection state and act according to the changes."""
+        if self.ssh_connection_state == SshConnectionState.waiting_for_input:
```

Comment:
[Enums should be compared by identity](https://docs.python.org/3.7/library/enum.html#using-automatic-values).